# Algorithm Design and Recursion

## 13.1 Searching

### 13.1.2 Stratergy 1: Linear Search

In [42]:
def search(x, nums):
    
    for i in range(len(nums)):
        if nums[i] == x:     # Item found, return the index value
            return i
        
    return -1               # loop finished, item was not in list

In [43]:
search(4,[3, 1, 4, 2, 5])

2

In [44]:
search(7,[3, 1, 4, 2, 5])

-1

### 13.1.3 Stratergy 1: Binary Search

In [45]:
def search(x, nums):
   
    low = 0
    high = len(nums) - 1
    
    while low <= high:        # There is still a range to search
        mid = (low+high)//2   # position of middle item
        item = nums[mid]
        
        if x == item:         # Found it! Return the index
            return mid
    
        elif x < item:        # x is in lower half of range
            high = mid - 1    #     Move top marker down
        
        else:                 # x is in upper half
            low = mid + 1     #     Move bottom marker up
    
    return -1                 # no range left to search, x is not there

In [46]:
search(4,[3, 1, 4, 2, 5])

2

In [47]:
search(7,[3, 1, 4, 2, 5])

-1

## 13.2 Recursive Problem Solving

In [ ]:
Algorithm: binarySearch -- search for x in numbs[low]...nums[high]

mid = (low+high)

if low > high
    x is not in nums
    
elif x < nums[mid]
    perform binary search for x in nums[low]...nums[mid-1]
    
else
    perform binary search for x in numbs[mid+1]...nums[high]

### 13.2.2 Recursive Functions

In [7]:
def fact(n):
    
    if n == 0:
        return 1
    
    else:
        return n * fact(n-1)

In [8]:
fact(4)

24

In [9]:
fact(10)

3628800

### 13.2.3 Example: String Reversal

In [11]:
def reverse(s):
    return reverse(s[1:]) + s[0]

In [12]:
reverse("Hello")

RecursionError: maximum recursion depth exceeded

In [13]:
def reverse(s):
    if s == "":
        return s
    else:
        return reverse(s[1:]) + s[0]

In [14]:
reverse("Hello")

'olleH'

### 13.2.4 Example: Anagrams

Come back to, as it's a little tricky.

In [15]:
def anagrams(s):
    
    if s == "":
        return [s]
    
    else:
        ans = []
        for w in anagrams(s[1:]):
            for pos in range(len(w) + 1):
                ans.append(w[:pos] + s[0] + w[pos:])
        
        return ans

In [48]:
s = "abs"

In [49]:
s[1:]

'bs'

In [55]:
s[:1]

'a'

In [56]:
s[0]

'a'

In [52]:
len(s)

3

In [16]:
anagrams("abs")

['abs', 'bas', 'bsa', 'asb', 'sab', 'sba']

### 13.2.5 Example: Fast Exponentiation

In [18]:
def loopPower(a, n):
    
    ans = 1
    
    for i in range(n):
        ans = ans * a
    
    return ans

In [57]:
loopPower(2,5)

32

In [21]:
def recPower(a, n):
    
    # raises a to the int power n
    if n == 0:
        return 1
    
    else:
        
        factor = recPower(a, n//2)
        
        if n%2 == 0:                    # n is even
            return factor * factor
        
        else:                          # n is odd
            return factor * factor * a

In [58]:
recPower(2,5)

32

In [59]:
recPower(2,0)

1

In [60]:
recPower(2,6)

64

In [94]:
recPower(3,6)

729

### 13.2.6 Example: Binary Search

In [23]:
def recBinSearch(x, nums, low, high):
    
    if low > high:              # No place left to look, return -1
        return -1
    
    mid = (low + high) // 2
    item = nums[mid]
    
    if item == x:               # Found it! Return the index
        return mid
    
    elif x < item:              # Look in lower half
        return recBinSearch(x, nums, low, mid-1)
    
    else:                       # Look in upper half
        return recBinSearch(x, nums, mid+1, high)

In [24]:
def search(x, nums):
    return recBinSearch(x, nums, 0, len(nums)-1)

### 13.2.7 Recursion vs. Iteration

In [25]:
def loopfib(n):
    # returns the nth Fibonacci number
    
    curr = 1
    prev = 1
    
    for i in range(n-2):
        curr, prev = curr + prev, curr
    
    return curr

In [26]:
def fib(n):
    
    if n < 3:
        return 1
    
    else:
        return fib(n-1) + fib(n02)

## 13.3 Sorting Algorithm

### 13.3.1 Naive Sorting: Selection Sort

In [ ]:
nums[0] = nums[10]

In [ ]:
nums[0], nums[10] = nums[10], nums[0]

In [28]:
def selSort(nums):
    # sort numbs into ascending order
    
    n = len(nums)
    
    # For each position in the list (except the very last)
    for bottom in range(n-1):
        # Find the smaller item in nums[bottom]..nums[n-1]
        
        mp = bottom                       # bottom is smallest initially
        for i in range(bottom + 1, n):    # look at each position
            if nums[i] < nums[mp]:        # this one is smaller
                mp = i                    #      remember its index
        
        # swap smallest item to the bottom
        nums[bottom], nums[mp] = nums[mp], nums[bottom]

In [61]:
nums = [2,3,4,5,5,6,7,7,75,4,3,3,34,354,546]

In [70]:
selSort(nums)
print(nums)

[2, 3, 3, 3, 4, 4, 5, 5, 6, 7, 7, 34, 75, 354, 546]


### 13.3.2 Divide and Conquer: Merge Sort

Algorithm : merge sort nums
    
split nums into two halves

sort the first half

sort the second half

merge the two sorted halves back into nums

In [30]:
def merge(lst1, lst2, lst3):
    # merge sorted lists lst1 and lst2 into lst3
    
    # these indexes keep track of current position in each list
    i1, i2, i3 = 0, 0, 0    # All start at the front
    n1, n2 = len(lst1), len(lst2)
    
    # Loop while both lst1 and lst2 have more items
    while i1 < n1 and i2 < n2:
        if lst1[i1] < lst2[i2]:   # top of lst1 is smaller
            lst3[i3] = lst1[i1]   #    copy it into current spot in lst 3
            i1 = i1 + 1
        
        else:                     # top of lst2 is smaller
            lst3[i3] = lst2[i2]   #    copy it into current spot in lst3
            i2 = i2 + 1
        
        i3 = i3 + 1               # item added to lst3, update position
        
    # Here either lst1 or lst2 is done. One of the following loops will
    # execute to finish up the merge.
    
    # Copy remaining items (if any) from lst1
    while i1 < n1:
        lst3[i3] = lst1[i1]
        i1 = i1 + 1
        i3 = i3 + 1
        
    # Copy remaining items (if any) from lst2
    while i2 < n2:
        lst3[i3] = lst2[i2]
        i2 = i2 + 1
        i3 = i3 + 1

In [71]:
nums1 =  [2,3,4,5,5,6,7,7,75,4,3,3,34,354,546]

In [72]:
nums2 = [435,435,23,54,76,98,67,234,23,654,76]

In [73]:
nums3 = [456,876,546,345,243,645,543,465,576,8]

In [75]:
merge(nums1, nums2, nums3)

IndexError: list assignment index out of range

if len(nums) > 1:
    
    split nums into two halves
    
    mergeSort the first half
    
    mergeSort the second half
    
    Merge the two sorted halves into nums

In [31]:
def mergeSort(nums):
    # Put items of nums in ascending order
    
    n = len(nums)
    
    # Do nothing if nums contains 0 or 1 item
    
    if n > 1:
        # split into two sublists
        
        m = n // 2
        nums1, nums2, = nums[:m], nums[m:]
        
        # recursively sort each piece
        mergeSort(nums1)
        mergeSort(nums2)
        
        # merge the sorted pieces back into original list
        merge(nums1, nums2, nums)

In [77]:
mergeSort(nums1)
print(nums1)

[2, 3, 3, 3, 4, 4, 5, 5, 6, 7, 7, 34, 75, 354, 546]


## 13.4 Hard Problems

### 13.4.1 Tower of Hanoi

Algorithm: move n-disk tower from source to destination via resting place

move n-1 disk tower from source to resting place

move 1 disk tower from source to destination

move n-1 disk tower from resting place to destination

In [90]:
def moveTower(n, source, dest, temp):
    
    count = 0
    
    if n == 1:
        print("Move disk from", source, "to", dest + ".")
    
    else:
        moveTower(n - 1, source, temp, dest)
        moveTower(1, source, dest, temp)
        moveTower(n - 1, temp, dest, source)

In [91]:
def hanoi(n):
    moveTower(n, "A", "C", "B")

In [92]:
hanoi(3)

Move disk from A to C.
Move disk from A to B.
Move disk from C to B.
Move disk from A to C.
Move disk from B to A.
Move disk from B to C.
Move disk from A to C.


In [83]:
hanoi(5)

Move disk from A to C.
Move disk from A to B.
Move disk from C to B.
Move disk from A to C.
Move disk from B to A.
Move disk from B to C.
Move disk from A to C.
Move disk from A to B.
Move disk from C to B.
Move disk from C to A.
Move disk from B to A.
Move disk from C to B.
Move disk from A to C.
Move disk from A to B.
Move disk from C to B.
Move disk from A to C.
Move disk from B to A.
Move disk from B to C.
Move disk from A to C.
Move disk from B to A.
Move disk from C to B.
Move disk from C to A.
Move disk from B to A.
Move disk from B to C.
Move disk from A to C.
Move disk from A to B.
Move disk from C to B.
Move disk from A to C.
Move disk from B to A.
Move disk from B to C.
Move disk from A to C.


### 13.4.2 The Halting Problem

In [ ]:
# turing.py

def terminates(program, inputData):
    # prgoram and inputData are both strings
    # Returns true if program would halt when run with inputData
    #    as its input.
    
def main():
    # Read a proogram from standard input
    lines = []
    print("Type in a program (type 'done' to quit).")
    line = input("")
    
    while line != "done":
        lines.append(line)
        line = input("")
    testProg = "\n".join(lines)
    
    # If program halts on itself as input, go into an infinite loop
    if terminates(testProg, testProg):
        whie True:
            pass       # a pass statement does nothing
        
main()

## 13.6 Exercises
### Review Questions
#### True/ False

1. Linear search requires a number of steps proportional to the size of the list being searched.

True.

Linear search scans the collection from start to end and requires time linearly proportional to the size of the collection.

2. The Python operator in performs a binary search.

False.

The Python Operator uses a Linear search.

A Linear search scans the collectection from start to end and requires time linearly proportional to the size of the collection.

3. Binary search is an n log n algorithm.

False.

log2 n

Binary search is an example of a log time algorithm. The amount of time it takes to solve a given problem grows as the log of the problem size. In the case of binary search, each additional iteration doubles the size of the problem that we can solve.

4. The number of times n can be divided by 2 is exp(n).

False.



5. All proper recursive definitions must have exactly one non-recursive base case.

False.

A definition or function is recursive if it refers to itself. To be well-founded, a recursive definition must meet two properties:

    1. There must be one or more base cases that require no recursion.
    
    2. All chains of recursion must eventually reach a base year.

6. A sequence can be viewed as a recursive data collection.

True.

Sequences can be considered recursive structures containing a first item followed by a sequence. Recursive functions can be written following this approach.

7. A word of length n has n! anagrams

True.

8. Loops are more general than recursion.

False.

Recursion is more general than iteration. Choosing between recursion and looping involves the considerations of effeciency and elegance.

9. Merge sort is an example of an n log n algorithm.

True.

Merge sort is a divide and conquer algorithm that can sort a collection in nlogn time.

10. Exponential algorithms are generally considered intractable.

True.

Problems that are solvable in theory but not in practice are called intractable. The solution to the famous Tower of Hanoi can be expressed as a simple recursive algorithm, but the algorithm is intractable.

#### Multiple Choice

1. Which algorithm requires time directly proportional to the size of the input?

    a. linear search
    
        Linear search scans the collection from start to end and requires time linearly proportional to the size of the collection.
    
    b. binary search
    
        Binary search is an example of a divide and conquer approach to algorithm development. Divide and conquer often yields effecient solutions.
    
    c. merge sort
    
        An effecient divide and conquer sorting algorithm.
        
            Example: splitting a deck then merging it in the end.
    
    d. selection sort
    
        An n squared-time sorting algorithm.
        
            Example: goes through a deck one by one from smallest to largerst.

My Answer. A. Linear search

2. Approximately how many iterations will binary search need to find a value in a list of 512 items?

    a. 512
    
    b. 256
    
    c. 9
    
    d. 3

log2 n

My answer.

C. 9

3. Recursions on sequences often use this as a base case:

    a. 0
    
    b. 1
    
    c. an empty sequence
    
    d. None

My answer:
    
b. 1

4. An infinite recursion will result in

    a. a program at "hangs"
    
    b. a broken computer
    
    c. a reboot
    
    d. a run-time exception

My answer.

d. a run-time exception.

If a recursion never reaches a base case, it goes on making recursive calls forever, and the program never terminates.

5. The recursive Fibonacci function is ineffecient because

    a. It does many repeated computations
    
    b. recursion is inherently ineffecient compared to iteration
    
    c. calculating Fibonacci numbers is intractable
    
    d. fibbing is morally wrong

My Answer.

a. it does mant repeated computations.

The recursive function is only useful only upto around 30.

6. Which is a quadratic time algorithm?

    a. Linear search
    
    b. Binary search
    
    c. Tower of hanoi
    
    d. Selection sort

My Answer.

d. Selection sort (n^2)

7. The process of combining two sorted sequences is called

    a. Sorting
    
    b. Shuffling
    
    c. Dovetailing
    
    d. Merging

My Answer.

d. Merging

8. Recursion is related to the mathematical technique called

    a. Looping
    
    b. Sequencing
    
    c. Induction
    
    d. Contradiction

My Answer. 

b. Sequencing.

Sequences can be considered recursive structures containing a first item followed by a sequence. Recursive functions can be written following this approach.

9. How many steps would be needed to solve the Tower of Hanoi for a tower of size 5?

    a. 5
    
    b. 10
    
    c. 25
    
    d. 31

My answer.

d. 31

10. Which of the following is not true of the halting problem?

    a. It was studied by Alan Turing.
    
    b. It is harder than intractable.
    
    c. Someday a clever algorithm may be found to solve it.
    
    d. It involves a program that analyzes other programs.

My answer.

b. It is harder than intractable.

#### Discussion

1. Place these algorithm classes in order from fastest to slowest: nlogn, n, n^2, log n, 2^n.

    1. logn


    2. nlogn


    3. n


    4. 2^n
    
    
    5. n^2

2. In your own words, explain the two rules that a proper recursive definition or function must follow.

A definition of function is recursive if it refers to itself. To be well-founded, a recursive definition must meet two properties.

    1. There must be one or more base cases that require no recursion.
    
    
    2. All chains of recursion must eventually reach a base case.

3. What is the exact result of anagram ("foo")?

oof

4. Trace recPower(3,6) and figure out exactly how many multiplications it performs.

729

5. Why are divide-and-conquer algorithms often very effecient?

Divide-and-conquer algorithms naturally tend to make efficient use of memory caches. The reason is that once a sub-problem is small enough, it and all its sub-problems can, in principle, be solved within the cache, without accessing the slower main memory.

#### Programming Exercises

1. Modify the recursive Fibonacci program given in this chapter so that is prints tracing information. Specifically, have the function print a message when it is called and when it returns. For example, the out should should contain lines like these:

    Computing fib(4)
    ...
    Leaving fib(4) returning 3
    
    Use your modified version of fib to compute fib(10) and count how many times fib(3) is computed in the process.

In [98]:
def moveTower(n, source, dest, temp):
    
    count = 0
    
    if n == 1:
        print("Move disk from", source, "to", dest + ".")
    
    else:
        moveTower(n - 1, source, temp, dest)
        moveTower(1, source, dest, temp)
        moveTower(n - 1, temp, dest, source)

In [99]:
def hanoi(n):
    moveTower(n, "A", "C", "B")

In [100]:
hanoi(3)

Move disk from A to C.
Move disk from A to B.
Move disk from C to B.
Move disk from A to C.
Move disk from B to A.
Move disk from B to C.
Move disk from A to C.


In [127]:
def fib(n):
    print("Computing fib({0})".format(n))
    
    if n < 3:
        return 1
    
    else:
        print("Leaving fib({0}), returning fib({1})".format(n, fib(n-1) + fib(n-2)))
        return fib(n-1) + fib(n-2)

In [130]:
fib(4)

Computing fib(4)
Computing fib(3)
Computing fib(2)
Computing fib(1)
Leaving fib(3), returning fib(2)
Computing fib(2)
Computing fib(1)
Computing fib(2)
Leaving fib(4), returning fib(3)
Computing fib(3)
Computing fib(2)
Computing fib(1)
Leaving fib(3), returning fib(2)
Computing fib(2)
Computing fib(1)
Computing fib(2)


3

3. A palindrome is a sentence that contains the same sequence of letters reading either forwards or backwards. 

    Check the textbook for more information.

In [133]:
def palindrome(xStr, firstRun=True):
    if firstRun == True:
        for ch in '<}{ ][>.?/:;\|+=-_),(*&^%$#@!~`':
            xStr.replace(ch, "")
        xStr = xStr.lower()
    
    if len(xStr) < 2:
        return True
    
    elif xStr[0] == xStr[-1]:
        xStr = xStr[1:-1]
        palindrome(xStr, firstRun=False)
        return True
    
    else:
        return False
        
def main():
    xStr = input("Type a phrase to check for palindrome status....")
    
    if palindrome(xStr) is True:
        print("The phrase is a palindrome.")
    else:
        print("The phrase is not a palindrome.")
        
main()

Type a phrase to check for palindrome status....a man, a plan, a canal, Panama!
The phrase is not a palindrome.


4. Write and test a recursive function max to find the largest number in a list. The max is the larger of the first item and the max of all other other items.

7. Write both an iterative and recursive function to compute combinations and compare the effeciency of your two solution.

    Hint: when k = 1, cnk = n and when n < k, cnk = 0.
    
    Check the textbook for more information.

8. Implement this algorithm with a Turtle classthat contains instance variables location (a Point) and Direction (a float) and methods such as moveTo(somePoint), draw(length), and turn(degrees).

    If you maintain direction as an angle in radians, the point you are going to can easily be computed from your current location.
    
    Just use dx = length * cos(direction) and dy = length * sin(direction).

9. Using an approach similar to the previous exercise, write a program that draws a C-curve. Hint: your turtle will do the following:

    Check the textbook for more information.